src/treemodel/composer/modifyselection.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Position from '../position.js';
+import TreeWalker from '../treewalker.js';
+import Range from '../range.js';
+
+/**
+ * Modifies the selection. Currently the supported modifications are:
+ *
+ * * Extending. The selection focus is moved in the specified `options.direction` with a step specified in `options.unit`
+ * (defaults to `'CHARACTER'`, other values are not not yet supported).
+ * Note: if you extend a forward selection in a backward direction you will in fact shrink it.
+ *
+ * @method engine.treeModel.composer.modifySelection
+ * @param {engine.treeModel.Selection} The selection to modify.
+ * @param {Object} [options]
+ * @param {'FORWARD'|'BACKWARD'} [options.direction='FORWARD'] The direction in which the selection should be modified.
+ */
+export default function modifySelection( selection, options ) {
+	if ( !options ) {
+		options = {};
+	}
+
+	const isForward = options.direction != 'BACKWARD';
+
+	const focus = selection.focus;
+	const walker = new TreeWalker( {
+		boundaries: getSearchRange( focus, isForward ),
+		singleCharacters: true
+	} );
+
+	const items = Array.from( walker );
```

Comment:
I am not sure if this is efficient, but I suppose that `modifySelection` won't be fired that often so I guess it's okay.